worlds/sa2b/__init__.py

```diff
@@ -170,20 +188,50 @@ def generate_basic(self):
                     current_gate = self.world.number_of_level_gates[self.player].value
                 else:
                     current_gate_emblems = max(
-                        math.floor(total_emblem_count * math.pow(total_levels_added / 30.0, 2.0)), current_gate)
+                        math.floor(total_emblem_count * math.pow(total_levels_added / 30.0, 2.0) * gate_cost_mult), current_gate)
                     gates.append(LevelGate(current_gate_emblems))
+                    self.gate_costs[current_gate] = current_gate_emblems
                 levels_added_to_gate = 0
 
         self.region_emblem_map = dict(zip(shuffled_region_list, emblem_requirement_list))
 
-        connect_regions(self.world, self.player, gates, self.emblems_for_cannons_core)
+        connect_regions(self.world, self.player, gates, self.emblems_for_cannons_core, self.gate_bosses)
 
         max_required_emblems = max(max(emblem_requirement_list), self.emblems_for_cannons_core)
         itempool += [self.create_item(ItemName.emblem)] * max_required_emblems
-        itempool += [self.create_item(ItemName.emblem, True)] * (total_emblem_count - max_required_emblems)
+
+        non_required_emblems = (total_emblem_count - max_required_emblems)
+        junk_count = math.floor(non_required_emblems * (self.world.junk_fill_percentage[self.player].value / 100.0))
+        itempool += [self.create_item(ItemName.emblem, True)] * (non_required_emblems - junk_count)
+
+        # Carve Traps out of junk_count
+        trap_weights = []
+        trap_weights += ([ItemName.omochao_trap] * self.world.omochao_trap_weight[self.player].value)
+        trap_weights += ([ItemName.timestop_trap] * self.world.timestop_trap_weight[self.player].value)
+        trap_weights += ([ItemName.confuse_trap] * self.world.confusion_trap_weight[self.player].value)
+        trap_weights += ([ItemName.tiny_trap] * self.world.tiny_trap_weight[self.player].value)
+
+        junk_count += extra_junk_count
+        trap_count = 0 if (len(trap_weights) == 0) else math.ceil(junk_count * (self.world.trap_fill_percentage[self.player].value / 100.0))
+        junk_count -= trap_count
+
+        junk_pool = []
+        for i in range(junk_count):
+            junk_item = self.world.random.choice(list(junk_table.keys()))
```

Comment:
running the conversion to list on every iteration is pretty expensive; you create a fresh list and then free the memory again every time.